### ISSUE TITLE
Use ATSynEdit instead of SynEdit

### ISSUE THREAD

**Issue by BBasile** [BBasile/Coedit#236]
This change is motivated by many small problems in  SynEdit and some nice features of AtSynEdit (like the minimap, multi carets, etc)

Much work to do. Branch where things happen is https://github.com/BBasile/Coedit/tree/atsynedit.
Progresss will be tracked here https://github.com/BBasile/Coedit/projects/2.
Success is not guaranteed though.

ping @Alexey-T. I may bother you a bit in the next weeks.

**Comment by Alexey-T** [comment]
Okay, bother me, I will read.

**Comment by BBasile** [comment]
I've just seen something that i dont like: https://github.com/Alexey-T/ATSynEdit_Ex/blob/master/atsynedit_ex/atsynedit_form_complete.pas. It's a separate package.
The form only fundamentaly need your controls for this: https://github.com/Alexey-T/ATSynEdit_Ex/blob/master/atsynedit_ex/atsynedit_form_complete.pas#L49.

Maybe you can use generics here: In your Ex package you specialize for you control and in ATSynEdit you specialize for a the LCL equivalent. What do you think ?

**Comment by BBasile** [comment]
Anyway that's not urgent. I may also fork and start a slightly different version.

**Comment by Alexey-T** [comment]
I think _form_complete must use always ATListbox (package ATFlatControls). for LCL listbox, new form neeed.

**Comment by BBasile** [comment]
What do you mean by "new form" needed ? I need to be sure of how to do the completion at the end of the week. Why ? ATM I've done the most simple things. Next i can do all the custom commands that have to be undo / redo able but after that it's a big pack: highlighter, options. I want to be sure that i'll use ATSynEdit before starting the "big pack of work".

**Comment by Alexey-T** [comment]
I mean, atsynedit-form-complete.pas should be forked and in fork you can use Listbox, not ATListbox. Is it ok.

**Comment by BBasile** [comment]
understood.

**Comment by Alexey-T** [comment]
Ok, IMHO change to listbox will be easy... only needed owner draw

**Comment by BBasile** [comment]
That's nice if so. Your first comment (2 days ago) was not clear and i thought that i would have code a full new visual component. I'll continue to work tomorrow.

By the way, if you want to participate you can accept the invitation i sent and then you'll have the right to push directly in my repo. If you do so, just don't forget to push to the feature branch made for ATSynEdit.

**Comment by BBasile** [comment]
I've started the custom commands today. It doesn't work well for now. I don't know why but the tokens i get after passing the strings text as UTF8 ends with an invalid token while it was not the case before. This breaks the method used to auto close the curly braces.

I also miss a command used to goto column and line XY, and that's undo / redo able.

**Comment by BBasile** [comment]
curly brace fixed but i reach very easily a FG list error due to undo redo like a mad man.

**Comment by Alexey-T** [comment]
Braces ok now, good
How to repeat list index error, can you post new issue, with steps?

**Comment by Alexey-T** [comment]
Or repeat it in CudaText app?

**Comment by BBasile** [comment]
When i disable UndoGroup of auto close custom command the bug disapears. Does `DoCommand(cCommand_Undo);` makes grouped undo by default ?

**Comment by BBasile** [comment]
I think that the answer is no. Then when i partially undo the custom command and add more changes there's a mess with the groups

**Comment by Alexey-T** [comment]
It is OptUndoGrouped seems?

**Comment by Alexey-T** [comment]
What is issue with undo?

**Comment by BBasile** [comment]
The `procedure Undo(AGrouped: boolean);` is certainly called by `DoCommand(cCommand_Undo);`.
Is it call with `AGrouped` set to `true` or false ? I suspect that this is the cause of the error i get since when don't use `Strings.BeginUndoGroup;` and `Strings.EndUndoGroup;` the error doesn't happen anymore.

**Comment by Alexey-T** [comment]
It’s called with val of OptUndoGrouped.

**Comment by Alexey-T** [comment]
note:
Update(true/false) - need if you called internals, eg changes Carets or Strings.
no need - if you only called DoCommand(). Docommand does it.

**Comment by BBasile** [comment]
Yeah i've seen the wiki change. This is actually the opposite of what you told me at the beginning. I can revert changes now.

**Comment by Alexey-T** [comment]
ops, I thought you changed Strings direct.

**Comment by Alexey-T** [comment]
(to not forget). I posted 2 comments inside last patch, inline in code. maybe you dont see.
